### DISCUSSION TITLE
Questions on how to use e3nn for ffnn

### DISCUSSION THREAD

**Discussion by jiachenguoNU** [e3nn/e3nn#371] [Q&A]
Hi, thanks for providing this tool for equivariant neural networks. I've been trying to understand how this works and how to implement this framework to my problem. But after reading the tutorials and API, I'm still unclear on how to use this tool to solve my problem.

In the problem I try to solve, the input is a 2nd-order symmetric tensor and a vector, the output is a 2nd-order symmetric tensor. I want to learn such a function using equivariant neural network that is equivariant to frame rotation. Since there is no graph structure here, I guess I need to use the FullyConnectedNet in the package. But there is no example of this. Could you please give me some hint on how to solve this problem using e3nn?

Thanks a lot!!

**Comment by Linux-cpp-lisp** [comment]
`FullyConnectedNet` is only for scalars... you could try to build something analogous by alternating `o3.Linear` layers and one of the equivariant nonlinearities, but that is not very expressive since there are no interactions between various irreps... as far as I know, how to build an expressive network on tensors alone without graph structure from these primitives is not a solved problem (and is one I've been curious about). The exception would, I think, be if you used the [`nn.S2Activation`](https://docs.e3nn.org/en/stable/api/nn/nn_s2act.html), which does induce some interactions between different irreps... someone else should probably confirm tho.

**Reply by jiachenguoNU** [reply]
Thanks for your reply!! Since I'm only starting to learn equivariant neural networks, my questions may sound very naive.  

Since my input is a vector along with a 2nd-order tensor, can I use the vector as the geometry and 2nd order tensor as the feature on that geometry? (this is what I've learned from the poster). The output is another 2nd-order tensor. Many thanks!

**Reply by Linux-cpp-lisp** [reply]
@jiachenguoNU ... probably that would work? You mean making a fake "graph" with two points displaced by your vector, and using the features on the central atom as your prediction? It's not exactly elegant but yes that would work in the sense that it would predict something symmetrically correct.

**Reply by jiachenguoNU** [reply]
Yes. That's what I'm thinking of
